src/toil/lib/threading.py

```diff
@@ -335,9 +333,222 @@ def process_name_exists(workDir, name):
+                pass
+            
+            # Make ourselves a file in it and lock it to prove we are alive.
+            self.lockfileFD, self.lockfileName = tempfile.mkstemp(dir=self.lockfileDir)
+            # Nobody can see it yet, so lock it right away
+            fcntl.lockf(self.lockfileFD, fcntl.LOCK_EX)
+            
+            # Now we're properly in, so release the global mutex
+            
+        log.debug('Now in arena %s', self.lockfileDir)
+        
+    def leave(self):
+        """
+        This process is leaving the arena. If this process happens to be the
+        last process standing, yields something, with other processes blocked
+        from joining the arena until the loop body completes and the process
+        has finished leaving. Otherwise, does not yield anything.
+        
+        Should be used in a loop:
+        
+            for _ in arena.leave():
+                # If we get here, we were the last process. Do the cleanup
+                pass
+        """
+        
+        # Make sure we're in it to start.
+        assert self.lockfileName is not None
+        assert self.lockfileFD is not None
+        
+        log.debug('Leaving arena %s', self.lockfileDir)
+        
+        with global_mutex(self.workDir, self.mutex):
+            # Now nobody else should also be trying to join or leave.
+            
+            # Take ourselves out.
+            try:
+                os.unlink(self.lockfileName)
+            except:
+                pass
+            self.lockfileName = None
+            fcntl.lockf(self.lockfileFD, fcntl.LOCK_UN)
+            os.close(self.lockfileFD)
+            self.lockfileFD = None
+            
+            for item in os.listdir(self.lockfileDir):
+                # There is someone claiming to be here. Are they alive?
+                full_path = os.path.join(self.lockfileDir, item)
+                
+                fd = os.open(full_path, os.O_RDONLY)
+                try:
+                    fcntl.lockf(fd, fcntl.LOCK_SH | fcntl.LOCK_NB)
+                except IOError as e:
+                    # Could not lock. It's alive!
+                    break
+                else:
+                    # Could lock. Process is dead.
+                    try:
+                        os.remove(full_path)
+                    except:
+                        pass
+                    fcntl.lockf(fd, fcntl.LOCK_UN)
+                    # Continue with the loop normally.
+            else:
+                # Nothing alive was found. Nobody will come in while we hold
+                # the global mutex, so we are the Last Process Standing.
+                log.debug('We are the Last Process Standing in arena %s', self.lockfileDir)
+                yield True
+                
+                try:
+                    # Delete the arena directory so as to leave nothing behind.
+                    os.rmdir(self.lockfileDir)
+                except:
```

Comment:
I suppose I'll ask a similar question here, what happens if this has an error and cannot delete the lock dir?

Is this worth posting a debug log message when this occurs?

Reply:
Here it might be sensible to have a warning, because we don't expect anything to still be there, and if something made it in, then somebody didn't respect the locks.